### PR TITLE
[Merged by Bors] - feat: port Data.Real.CauSeq

### PR DESCRIPTION
**Pull request by mcdoll**
- [x] depends on: #1398

### CODE REVIEW

File: Mathlib/Algebra/Group/Pi.lean [hunk 1]
```diff
@@ -93,7 +93,11 @@ instance monoid [∀ i, Monoid <| f i] : Monoid (∀ i : I, f i) :=
 #align pi.add_monoid Pi.addMonoid
 
 instance addMonoidWithOne [∀ i, AddMonoidWithOne <| f i] : AddMonoidWithOne (∀ i : I, f i) :=
-  { addMonoid with }
+  { addMonoid with
+    natCast := fun n _ => n
+    natCast_zero := funext fun _ => AddMonoidWithOne.natCast_zero
+    natCast_succ := fun n => funext fun _ => AddMonoidWithOne.natCast_succ n
+  }
```
Comment (dupuisf) [inline Mathlib/Algebra/Group/Pi.lean:100]:
```suggestion
    natCast_succ := fun n => funext fun _ => AddMonoidWithOne.natCast_succ n }
```

File: Mathlib/Algebra/Group/Pi.lean [hunk 2]
```diff
@@ -144,7 +148,11 @@ instance group [∀ i, Group <| f i] : Group (∀ i : I, f i) :=
 #align pi.add_group Pi.addGroup
 
 instance addGroupWithOne [∀ i, AddGroupWithOne <| f i] : AddGroupWithOne (∀ i : I, f i) :=
-  { addGroup with }
+  { addGroup, addMonoidWithOne with
+    intCast := fun z _ => z
+    intCast_ofNat := fun n => funext fun _ => AddGroupWithOne.intCast_ofNat n
+    intCast_negSucc := fun n => funext fun _ => AddGroupWithOne.intCast_negSucc n
+  }
```
Comment (dupuisf) [inline Mathlib/Algebra/Group/Pi.lean:155]:
```suggestion
    intCast_negSucc := fun n => funext fun _ => AddGroupWithOne.intCast_negSucc n }
```

File: Mathlib/Data/Real/CauSeq.lean [hunk 3]
```diff
@@ -0,0 +1,1028 @@
...
+* `IsCauSeq`: a predicate that says `f : ℕ → β` is Cauchy.
+* `CauSeq`: the type of Cauchy sequences valued in type `β` with respect to an absolute value
+  function `abv`.
+## Tags
+sequence, cauchy, abs val, absolute value
+-/
+
+
+--set_option autoImplicit false
+
```
Comment (dupuisf) [inline Mathlib/Data/Real/CauSeq.lean:35]:
```suggestion
```

File: Mathlib/Data/Real/CauSeq.lean [hunk 4]
```diff
@@ -0,0 +1,1028 @@
...
+open IsAbsoluteValue
+
+--variable {G α β : Type _}
+
```
Comment (dupuisf) [inline Mathlib/Data/Real/CauSeq.lean:39]:
```suggestion
```

File: Mathlib/Data/Real/CauSeq.lean [hunk 5]
```diff
@@ -0,0 +1,1028 @@
...
+instance {β : Type _} [CommRing β] {abv : β → α} [IsAbsoluteValue abv] : CommRing (CauSeq β abv) :=
+  { CauSeq.ring with
+    mul_comm := fun a b => ext $ fun n => by simp [mul_left_comm, mul_comm] }
+
+/-- `LimZero f` holds when `f` approaches 0. -/
+def LimZero {abv : β → α} (f : CauSeq β abv) : Prop :=
+  ∀ ε > 0, ∃ i, ∀ j ≥ i, abv (f j) < ε
+#align cau_seq.lim_zero CauSeq.LimZero
+
+theorem add_lim_zero {f g : CauSeq β abv} (hf : LimZero f) (hg : LimZero g) : LimZero (f + g)
```
Comment (dupuisf) [inline Mathlib/Data/Real/CauSeq.lean:436]:
```suggestion
theorem add_limZero {f g : CauSeq β abv} (hf : LimZero f) (hg : LimZero g) : LimZero (f + g)
```

File: Mathlib/Data/Real/CauSeq.lean [hunk 6]
```diff
@@ -0,0 +1,1028 @@
...
+  | ε, ε0 =>
+    (exists_forall_ge_and (hf _ <| half_pos ε0) (hg _ <| half_pos ε0)).imp fun i H j ij => by
+      let ⟨H₁, H₂⟩ := H _ ij
+      simpa [add_halves ε] using lt_of_le_of_lt (abv_add _ _) (add_lt_add H₁ H₂)
+#align cau_seq.add_lim_zero CauSeq.add_lim_zero
+
+theorem mul_lim_zero_right (f : CauSeq β abv) {g} (hg : LimZero g) : LimZero (f * g)
```
Comment (dupuisf) [inline Mathlib/Data/Real/CauSeq.lean:443]:
```suggestion
theorem mul_limZero_right (f : CauSeq β abv) {g} (hg : LimZero g) : LimZero (f * g)
```

Comment (dupuisf) [reply Mathlib/Data/Real/CauSeq.lean:443]: And the same throughout.

File: Mathlib/Data/Real/CauSeq.lean [hunk 7]
```diff
@@ -0,0 +1,1028 @@
...
+/-instance : AddGroupWithOne (CauSeq β abv) :=
+  { CauSeq.addGroup with
+    one := 1
+    natCast := (fun n => const n)
+    natCast_zero := congr_arg const Nat.cast_zero
+    natCast_succ := fun n => congr_arg const (Nat.cast_succ n)
+    intCast := (fun n => const n)
+    intCast_ofNat := fun n => congr_arg const (Int.cast_ofNat n)
+    intCast_negSucc := fun n => congr_arg const (Int.cast_negSucc n) }-/
+
```
Comment (dupuisf) [inline Mathlib/Data/Real/CauSeq.lean:403]: This should probably be deleted, no?

File: Mathlib/Data/Real/CauSeq.lean [hunk 8]
```diff
@@ -0,0 +1,1028 @@
...
+variable [Ring β] {abv : β → α}
+
+instance : CoeFun (CauSeq β abv) fun _ => ℕ → β :=
+  ⟨Subtype.val⟩
+
+-- Porting note: Remove coeFn theorem
+/-@[simp]
+theorem mk_to_fun (f) (hf : IsCauSeq abv f) : @coeFn (CauSeq β abv) _ _ ⟨f, hf⟩ = f :=
+  rfl
+#align cau_seq.mk_to_fun CauSeq.mk_to_fun-/
```
Comment (eric-wieser) [inline Mathlib/Data/Real/CauSeq.lean:154]: Should this be `noalign`?

File: Mathlib/Data/Real/CauSeq.lean [hunk 9]
```diff
@@ -0,0 +1,1028 @@
...
+#align cau_seq.const_sub CauSeq.const_sub
+
+section SMul
+
+variable [SMul G β] [IsScalarTower G β β]
+
+instance : SMul G (CauSeq β abv) :=
+  ⟨fun a f => (ofEq (const (a • (1 : β)) * f) (a • (f : ℕ → β))) fun _ => smul_one_mul _ _⟩
+
+@[simp] -- Porting note: Removed `norm_cast` attribute
```
Comment (eric-wieser) [inline Mathlib/Data/Real/CauSeq.lean:357]: Do we have an issue or zulip thread about this?

File: Mathlib/Data/Real/CauSeq.lean [hunk 10]
```diff
@@ -0,0 +1,1028 @@
...
+  change LimZero (f1 * g1 - f2 * g2)
+  convert add_lim_zero (mul_lim_zero_left g1 hf) (mul_lim_zero_right f2 hg) using 1
+  rw [mul_sub, sub_mul]
+  -- doesn't work with `rw`, but did in Lean 3
+  exact (sub_add_sub_cancel (f1*g1) (f2*g1) (f2*g2)).symm
+  -- was:
+  /-
+  simpa only [mul_sub, sub_mul, sub_add_sub_cancel] using
+    add_lim_zero (mul_lim_zero_left g1 hf) (mul_lim_zero_right f2 hg)
+  -/
```
Comment (eric-wieser) [inline Mathlib/Data/Real/CauSeq.lean:596]: This needs the words "porting note" somewhere.

File: Mathlib/Data/Real/CauSeq.lean [hunk 11]
```diff
@@ -0,0 +1,1017 @@
...
+#align rat_mul_continuous_lemma rat_mul_continuous_lemma
+
+theorem rat_inv_continuous_lemma {β : Type _} [DivisionRing β] (abv : β → α) [IsAbsoluteValue abv]
+    {ε K : α} (ε0 : 0 < ε) (K0 : 0 < K) :
+    ∃ δ > 0, ∀ {a b : β}, K ≤ abv a → K ≤ abv b → abv (a - b) < δ → abv (a⁻¹ - b⁻¹) < ε := by
+  refine' ⟨K * ε * K, mul_pos (mul_pos K0 ε0) K0, @fun a b ha hb h => _⟩
+  have a0 := K0.trans_le ha
+  have b0 := K0.trans_le hb
+  rw [inv_sub_inv' ((abv_pos abv).1 a0) ((abv_pos abv).1 b0)]
+  rw [@abv_mul _ _ _ _ abv _ _ b⁻¹] -- Porting note: mistake in `abv_mul`?
```
Comment (eric-wieser) [inline Mathlib/Data/Real/CauSeq.lean:85]: I think we should fix this now rather than working around it in every downstream port.

**Comment by eric-wieser** [comment]
I added a commit to compute the diff with mathport [here](https://github.com/leanprover-community/mathlib4/compare/3ba5e5fdf7665d12103b853e634d716553020f9e..459e2ed379a069952f927ea6511bebfd481176a9#diff-18f1434dcab7d38fad2c5144ab5f653c1eef55b9d1d49d20951e556af4ad0317L4).

The bulk of the difference seems to be mistakes in the porting of `abv_add` and similar lemmas. I think we should make a separate PR to fix those, then come back to this; otherwise we will be fixing every single file that references them.

File: Mathlib/Data/Real/CauSeq.lean (hunk 11 above)
Comment (mcdoll) [reply Mathlib/Data/Real/CauSeq.lean:85]: Made a PR for that: #1398

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#1398
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#1398~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/Data/Real/CauSeq.lean [hunk 12]
```diff
@@ -0,0 +1,1014 @@
...
+    f1 * g1 ≈ f2 * g2 := by
+  change LimZero (f1 * g1 - f2 * g2)
+  convert add_limZero (mul_limZero_left g1 hf) (mul_limZero_right f2 hg) using 1
+  rw [mul_sub, sub_mul]
+  -- Porting note: doesn't work with `rw`, but did in Lean 3
+  exact (sub_add_sub_cancel (f1*g1) (f2*g1) (f2*g2)).symm
+  -- Porting note: was
+  /-
+  simpa only [mul_sub, sub_mul, sub_add_sub_cancel] using
+    add_lim_zero (mul_lim_zero_left g1 hf) (mul_lim_zero_right f2 hg)
```
Comment (eric-wieser) [inline Mathlib/Data/Real/CauSeq.lean:581]: I think we should correct the names in this comment,
```suggestion
  simpa only [mul_sub, sub_mul, sub_add_sub_cancel] using
    add_limZero (mul_limZero_left g1 hf) (mul_limZero_right f2 hg)
```

**Comment by eric-wieser** [comment]
[here's](https://github.com/leanprover-community/mathlib4/compare/154f76b355130e6178b0322a7b682eff1f9e0cba..b67c547a34f7a3bd58ba283eba61589bb18e5e05) the new mathport diff.

File: Mathlib/Data/Real/CauSeq.lean [hunk 13]
```diff
@@ -0,0 +1,1011 @@
...
+#align is_cau_seq IsCauSeq
+
+namespace IsCauSeq
+
+variable [LinearOrderedField α] [Ring β] {abv : β → α} [IsAbsoluteValue abv] {f g : ℕ → β}
+
+-- see Note [nolint_ge]
+--@[nolint ge_or_gt] -- Porting note: restore attribute
+theorem cauchy₂ (hf : IsCauSeq abv f) {ε : α} (ε0 : 0 < ε) :
+    ∃ i, ∀ (j) (_ : j ≥ i) (k) (_ : k ≥ i), abv (f j - f k) < ε := by
```
Comment (eric-wieser) [inline Mathlib/Data/Real/CauSeq.lean:105]: @digama0, is the right thing to do with the
```
/- ./././Mathport/Syntax/Translate/Basic.lean:632:2: warning: expanding binder collection (j k «expr ≥ » i) -/
```
to just [remove the comment](https://github.com/leanprover-community/mathlib4/compare/154f76b355130e6178b0322a7b682eff1f9e0cba..mcdoll/data_real_causeq#diff-18f1434dcab7d38fad2c5144ab5f653c1eef55b9d1d49d20951e556af4ad0317L108)? Or should it have been left in place?

File: Mathlib/Data/Real/CauSeq.lean [hunk 14]
```diff
@@ -0,0 +1,1011 @@
...
+  Subtype.eq (funext h)
+#align cau_seq.ext CauSeq.ext
+
+theorem is_cau (f : CauSeq β abv) : IsCauSeq abv f :=
+  f.2
+#align cau_seq.is_cau CauSeq.is_cau
+
+theorem cauchy (f : CauSeq β abv) : ∀ {ε}, 0 < ε → ∃ i, ∀ j ≥ i, abv (f j - f i) < ε := by
+  have := f.2
+  rwa [IsCauSeq] at this
```
Comment (eric-wieser) [inline Mathlib/Data/Real/CauSeq.lean:161]: Did the [mathport output](https://github.com/leanprover-community/mathlib4/compare/154f76b355130e6178b0322a7b682eff1f9e0cba..mcdoll/data_real_causeq#diff-18f1434dcab7d38fad2c5144ab5f653c1eef55b9d1d49d20951e556af4ad0317L166-R161) not work here?

File: Mathlib/Data/Real/CauSeq.lean (hunk 13 above)
Comment (digama0) [reply Mathlib/Data/Real/CauSeq.lean:105]: I think you can use the bunder collection syntax here, i.e. `∃ i, ∀ j ≥ i,  ∀ k ≥ i, abv (f j - f k) < ε`

File: Mathlib/Data/Real/CauSeq.lean (hunk 14 above)
Comment (mcdoll) [reply Mathlib/Data/Real/CauSeq.lean:161]: It does not work.

Comment (mcdoll) [reply Mathlib/Data/Real/CauSeq.lean]: but `@f.2` does

**Review by Ruben-VandeVelde — APPROVED** [review]
lgtm

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**